### ISSUE TITLE
Saving a model with SparseCategoricalCrossentropy loss breaks subsequent training

### ISSUE THREAD

**Issue by drasmuss** [keras-team/keras#17511]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04 (WSL)
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.11.0
- Python version: 3.9
- Bazel version (if compiling from source): N/A
- GPU model and memory: N/A
- Exact command to reproduce: https://gist.github.com/drasmuss/62b129c02333da2ae01125daf50924a1

**Describe the problem**.

Calling `model.save` on a model with `SparseCategoricalCrossentropy` loss breaks the model somehow, so that subsequent calls to `model.fit` don't train successfully.

**Describe the current behavior**.

After calling `model.save`, subsequent `model.fit` training produces chance-level accuracy.

**Describe the expected behavior**.

The training should not be affected by calling `model.save`, accuracy should be the same as when we don't call `model.save`.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no): no
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing): N/A

**Standalone code to reproduce the issue**.

https://gist.github.com/drasmuss/62b129c02333da2ae01125daf50924a1

Note that this is just the standard MNIST convnet example from https://keras.io/examples/vision/mnist_convnet/, except I changed it to use `SparseCategoricalCrossentropy` loss and added the `model.save` call to trigger the bug.

**Source code / logs**.

Training results without `model.save`:
```
422/422 [==============================] - 13s 8ms/step - loss: 0.3842 - accuracy: 0.8864 - val_loss: 0.0826 - val_accuracy: 0.9773
```

Training results after calling `model.save`:
```
422/422 [==============================] - 10s 6ms/step - loss: 0.3632 - accuracy: 0.0998 - val_loss: 0.0864 - val_accuracy: 0.0978
```

**Comment by tilakrayal** [comment]
@sachinprasadhs,
I was able to reproduce the issue on tensorflow v2.9, v2.11 and nightly. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/50dba88b671242d634dd66a77b7dbddb/untitled913.ipynb).

**Comment by grasskin** [comment]
@nkovela1
